### ISSUE TITLE
Is this a bug somewhere?

### ISSUE THREAD

**Issue by audetto** [ihewitt/poolviewer#19]
http://pastebin.com/FbX630pG

This is my data.
Downloaded with the Windows app.

If you look at the 1st set on 04/10/2015 you can see in the summary that it took me

1:45 to do 75 M

This corresponds to a speed of 141 (per 100M)

But when you look at the split times (both here and in the windows app) you can only see 3 lanes totaling: 1:22 which is about my speed over 75M

I never do odd number of lanes, so I think it has lost one turn, but
it has lost time too!!!!

Should there be an option to ignore the total times and sum the split times?

**Comment by audetto** [comment]
Actually, I think it is more correct to add a lane with the time missing.

**Comment by audetto** [comment]
I've tested qttest and it never goes to line 172.

This is what it prints

        Len time 22.12, Strks 8
        Len time 30.25, Strks 12
        Len time 28.00, Strks 11
Set lengths 3, Time: 01:46

Which is missing 1 lap.

**Comment by ihewitt** [comment]
Can you run the latest version of qtest and pastebin the dumped output, and the matching lines from the windows csv. I'm getting the hang of decoding these now. :)
Also if you give the main app a go, the syncing is starting to work (although a bit fragile at the moment, occasionally I've needed to re-sync to get the serial to start).

**Comment by audetto** [comment]
probably missing a "git push"

**Comment by ihewitt** [comment]
It was late. :)
I'll do it when I get home.

**Comment by ihewitt** [comment]
pushed. :)

**Comment by audetto** [comment]
I've tried the new version and the output is exactly the same as the win app.
qtttest output is here:

http://pastebin.com/v3GLKmTH

either the watch tries to detect period when I do not move much,
and I cannot believe that my breaststroke is detected as "not swimming"

or for some reasons it misses the "last (?)" lanes?
all the other ones have correct times.

I think it is probably the first.

**Comment by ihewitt** [comment]
yes, there's no additional length data in that dump.
Not sure how you could arbitrarily allocate the additional time to an extra length though. How would you differentiate between resting and swimming? also you'd have to generate a theoretical number of strokes that that period too. Not sure that'd provide you with anything useful to look at. 
Might just have to resort to highlighting unusual sets and provide an edit/insert feature to do it manually.

**Comment by audetto** [comment]
I was going to add a new column to the set view called GAP to display missing (or extra) time.

Then as a separate pr, a way to turn this missing time into lanes.

**Comment by ihewitt** [comment]
The per set rest time should be getting populated correctly now, although I
need to go for a swim to get some more data to check. (had to reset watch)

**Comment by audetto** [comment]
What do you mean?
*Rest* is at the end of a set after one presses start. And this is correctly recorded and I can already see it (at least in the .csv generated by Windows). The app displays it properly.

*Missing* is the difference between sum of lanes time and set time, which I hardly consider rest, in all my cases these are missed lanes.

I am referring at the latter when I say I want to convert in lanes.

**Comment by audetto** [comment]
I've pushed the initial version of the editor to convert gap -> lanes.
One can only preview without being able to save the new set.

Probably the user needs to have a few more bits of info, but it is most of the implementation of my original idea.

It is not a push request yet, but the curious can see it at

https://github.com/audetto/poolviewer/commits/missing

**Comment by audetto** [comment]
It is almost ready,
I've pushed the latest version, I will rebase it when I propose the PR.

one outstanding issue is GUI updating and workout consistency.

I have noticed that the existing "delete set" button suffers from a similar issue.

1) the workout contains redundant fields about total distance, max min efficiency which are not updated when a set is removed.
2) if they were updated, the GUI would not follow as the call to "workoutSelected()" does not repaint the workouts. One can call the 3 functions called when a workout is removed, but then they change the selected (workout,set) causing a jump in the display

I am not trying to solve the particular case of editing a set, probably a more comprehensive solution is required, where everything is refreshed without changing selections.

The same, when sets change or are removed, the workout totals should be updated uniformly.

**Comment by audetto** [comment]
And should the user be allowed to discard changes (like delete workout or set)?
There is a backup now, but maybe one would like not to save if it does not look good.

**Comment by ihewitt** [comment]
Agreed on both points, the workout->data synchronisation needs tidying up now it's getting more complex. It started out quite simply, but the gui has evolved somewhat.
similarly an undo or cancel changes would be sensible.
